### PR TITLE
NIGH-152 Log Findings to CircleCI UI + Make Github Token Optional

### PR DESCRIPTION
**Pull request by alan20854**
implementing suggested improvements from circle rep call (allow for orb functionality even if users don't provide a github token, and always log finding messages to circle UI)

### CODE REVIEW

File: cmd/nightfalldlp/main.go [hunk 1]
```diff
@@ -91,9 +91,9 @@ func CreateDiffReviewerClient() (diffreviewer.DiffReviewer, error) {
 	case usingCircleCi():
 		githubToken, ok := os.LookupEnv(githubTokenEnvVar)
 		if !ok {
```
Comment (alan20854) [inline cmd/nightfalldlp/main.go]: or empty string

**Comment by alan20854** [comment]
make sure to test with https://github.com/nightfallai/nightfall_code_scanner/pull/41 merged in

File: cmd/nightfalldlp/main.go [hunk 2]
```diff
@@ -90,10 +90,10 @@ func CreateDiffReviewerClient() (diffreviewer.DiffReviewer, error) {
 		return github.NewAuthenticatedGithubService(githubToken), nil
 	case usingCircleCi():
 		githubToken, ok := os.LookupEnv(githubTokenEnvVar)
-		if !ok {
-			return nil, fmt.Errorf("could not find required %s environment variable", githubTokenEnvVar)
+		if !ok || githubToken == "" {
+			return circleci.NewUnauthenticatedCircleCiService(), nil
```
Comment (bradleymcallister97) [inline cmd/nightfalldlp/main.go]: It feels like we should do an info log if the github token is missing, to indicate that no findings will be directly posted to github, thoughts?

Comment (alan20854) [reply cmd/nightfalldlp/main.go]: sure, that makes sense

File: internal/clients/diffreviewer/circleci/circleci_service_test.go [hunk 3]
```diff
@@ -226,6 +226,51 @@ func (c *circleCiTestSuite) TestGetDiff() {
...
+		"testComment",
+		"/comments.txt",
+		tp.cs.PrDetails.CommitSha,
+		60,
+	)
+	emptyComments := []*diffreviewer.Comment{}
+
+	tests := []struct {
+		giveComments []*diffreviewer.Comment
+		wantError    error
```
Comment (bradleymcallister97) [inline internal/clients/diffreviewer/circleci/circleci_service_test.go]: This var is unused in the test

File: internal/clients/diffreviewer/circleci/circleci_service_test.go [hunk 4]
```diff
@@ -226,6 +226,51 @@ func (c *circleCiTestSuite) TestGetDiff() {
 	c.Equal(expectedFileDiffs, fileDiffs, "invalid fileDiff return value")
 }
 
+func (c *circleCiTestSuite) TestWriteCircleComments() {
+	tp := c.initTestParams()
+	ctrl := gomock.NewController(c.T())
+	defer ctrl.Finish()
+	testCircleService := &Service{
+		Logger: circlelogger.NewDefaultCircleLogger(),
```
Comment (bradleymcallister97) [inline internal/clients/diffreviewer/circleci/circleci_service_test.go]: Can we use a mock logger and write expects on the comments we expect to be written? Right now we are only validating that there is no error which doesn't really show that the comments are actually written

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 5]
```diff
@@ -54,14 +54,21 @@ type prDetails struct {
 	PrNumber  *int
 }
 
-// NewCircleCiService creates a new CircleCi service
-func NewCircleCiService(token string) diffreviewer.DiffReviewer {
+// NewAuthenticatedCircleCiService creates a new CircleCi service with an authenticated Github client
+func NewAuthenticatedCircleCiService(token string) diffreviewer.DiffReviewer {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: nit - can we update this naming to make it a little clearer that the only real difference is whether we write comments to github PR or not? Maybe like `NewCircleCIService` and `NewCircleCIServiceWithGHComments` or something?

File: internal/clients/diffreviewer/circleci/circleci_service_test.go [hunk 6]
```diff
@@ -226,6 +226,51 @@ func (c *circleCiTestSuite) TestGetDiff() {
...
+		{
+			giveComments: emptyComments,
+			wantError:    nil,
+			desc:         "no comments test",
+		},
+	}
+
+	for _, tt := range tests {
+		err := tp.cs.WriteComments(tt.giveComments)
+		c.NoError(err, fmt.Sprintf("Error writing comments for %s test", tt.desc))
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service_test.go]: yup about what brad said - we care specifically about what's being written, less so about whether or not there's an error (although obviously good to also assert that there isn't)

Comment (henkejosh) [reply internal/clients/diffreviewer/circleci/circleci_service_test.go]: also probably good to test each of the CircleServices - 1 assert that the corret mock GHComments calls are being made, the other that the logger is logging the appropriate info

File: internal/clients/diffreviewer/circleci/circleci_service_test.go [hunk 7]
```diff
@@ -226,15 +227,77 @@ func (c *circleCiTestSuite) TestGetDiff() {
 	c.Equal(expectedFileDiffs, fileDiffs, "invalid fileDiff return value")
 }
 
+func (c *circleCiTestSuite) TestWriteCircleComments() {
```
Comment (alan20854) [inline internal/clients/diffreviewer/circleci/circleci_service_test.go:228]: circle service without github client

File: internal/clients/diffreviewer/circleci/circleci_service_test.go [hunk 8]
```diff
@@ -385,6 +461,12 @@ func (c *circleCiTestSuite) TestWriteRepositoryComments() {
 				testCircleService.PrDetails.CommitSha,
 				gc,
 			)
+			mockLogger.EXPECT().Error(fmt.Sprintf(
+				"%s at %s on line %d",
+				gc.GetBody(),
+				gc.GetPath(),
+				tt.giveComments[index].LineNumber,
```
Comment (alan20854) [inline internal/clients/diffreviewer/circleci/circleci_service_test.go:466]: unlike `github.PullRequestComment`, `github.RepositoryComment` does not contain a line number field, although it is still applicable for the user to trace the comment from circle workflow to their github repo with a line number e.g.:
Screen Shot 2020-09-03 at 5 53 51 PM

**Review by bradleymcallister97 — APPROVED** [review]

**Review by henkejosh — APPROVED** [review]

**Pull request merged**